Make environment logging configurable

kalkwarf opened the pull request:
Resolves https://github.com/tuist/tuist/issues/YYY

### Short description 📝

Printing the ENV for every script phase really pollutes build logs. This PR allows users to disable that.

### Solution 📦

I've added an argument to `ExecutionAction` to control whether logging is enabled. I've defaulted it to `true` to preserve existing behavior, and to make this a non-breaking change.

### Implementation 👩‍💻👨‍💻

Extended the `ExecutionAction` struct, and made the initializer default the new value to true.

In the unit tests, left the pre-action alone, and set the post-action to use `false`.

Additionally introduced `XCTUnwrap` to clean up the optionals, primarily so the optional Bool test didn't need to go through contortions.

kwridan reviewed (commented):
Thanks for submitting this @kalkwarf 

Looks like there may be a few missing bits. The changes you made so far are the right ones to get the logic working internally, I'm assuming the goal was to expose this to the `Project.swift` manifests or perhaps something to apply globally e.g. an additional `generationOption` setting in `Config.swift`?

There are two `ExecutionAction` types

- `ProjectDescription.ExecutionAction` (public manifest types)
- `TuistCore.ExecutionAction` (internal model types)

They are mostly similar, however the internal model types expose more details which sometimes are omitted or exposed in a simpler manner to the public manifest types.

To expose it publicly the following components would need an update:

- `ProjectDescription.ExecutionAction` (public manifest type)
- `ExecutionAction+ManifestMapper`(component that converts public manifest type to the internal model types)
- Lastly the docs would need to be updated to highlight this new option and what it does

The alternative to apply it globally (slightly more involved but could be nicer overall to limit exposed properties):

-  Omit exposing it in `ProjectDescription.ExecutionAction`
- Introduce a new `generationOption` in `Config.swift` (e.g. `disableShowingEnvironmentInScriptBuildPhases`) 
- Introduce a new `TargetMapper` - `TargetActionEnvironmentMapper` 

```swift
class TargetActionEnvironmentMapper: TargetMapping {
   let showEnvVarsInLog: Bool
    
   fun map(target: Target) -> Target {
       // update the `showEnvVarsInLog` within all actions in this target
   }
}
```

- Wire it up in `ProjectMapperProvider`

```swift
class ProjectMapperProvider: ProjectMapperProviding {
    func mapper(config: Config) -> ProjectMapping {
        var mappers: [ProjectMapping] = []
        // ...
        if config.generationOptions.contains(. disableShowingEnvironmentInScriptBuildPhases) {
             mappers.append(
                 TargetProjectMapper(mapper: TargetActionEnvironmentMapper(showEnvVarsInLog: false))
             )
         }
   }
}
```

- Lastly the docs would need to be updated to highlight this new option and what it does


Hope this helps :)

kalkwarf commented:
Thank you for the awesome writeup explaining the missing functionality.

I've updated the PR to make this a `Config` level flag, added tests that show the flag getting toggled, and tweaked the documentation as well.

When I try running the build in Xcode, I get an error `Couldn't find ProjectDescription` which tracing through the code seems like the ProjectDescription.framework isn't built/visible. I don't _think_ that's anything I did?

Please let me know if there's anything else missing.

pepibumur commented:
@kalkwarf try rebasing from master 🙂. That's a known issue that @kwridan fixed and the fix is in master.

pepibumur commented:
@all-contributors please add @kalkwarf for ideas

kwridan commented:
Thanks for applying the updates @kalkwarf 👍 

As @pepibumur mentioned, we've landed a few fixes to re-enable debugging Tuist in Xcode #1786 and pushed #1798 to update the docs accordingly.

As for what's missing, apologies I missed 2 steps 

- As with `ExecutionAction`, there are two versions of `Config.swift` 
  - `ProjectDescription.Config` _(public manifest type)_
  - `TuistCore.Config` _(internal model type)_
  - As such we'll need to update  `ProjectDescription.Config` to also include the new option and the custom encoding / decoding to support it
- Update of `Config+ManifestMapper` to convert from `ProjectDescription.Config` > `TuistCore.Config` 

Feel free to create a new fixture in `fixtures` to help you test this out end-to-end with all the pieces in place (either via Xcode or the command line)

```swift
cd fixtures/<new-fixture-with-target-actions>
swift build
swift run tuist generate
```

kwridan reviewed (commented):
Thanks @kalkwarf it's nearly ready to ship 🚀! Were you able to get it up and running locally with the latest updates? 

The changes look good and work nicely - I manually tested this locally by making myself a dummy fixture and used the new config option.

![Screenshot 2020-09-22 at 08 39 08](https://user-images.githubusercontent.com/11914919/93855251-1af59580-fcaf-11ea-95dc-4385038a899d.png)


Some final bits _(thanks for your patience!)_:

- Looks like the latest merge from master may have gone a bit funny, the PR is showing unrelated changes included :( 
- You can run `bin/swiftformat Sources Tests` to ensure the same version that is used on CI formats the code

Minor suggestions:
- The `TargetActionEnvironmentMapper`/Tests could be in its own file in `TuistGenerator/Mappers` rather that part of the `TargetMapper` file in the core library
- `ConfigTests` can be extended to also include `disableShowEnvironmentVarsInScriptPhases`

kalkwarf commented:
> Were you able to get it up and running locally with the latest updates?

No, I'm really fighting Xcode & SPM.

- pull master
- run `open Package.swift`
- wait for the dependencies to download
- build `ProjectDescription`
- run `tuist` with the argument `init`
_this succeeds_
- run `tuist` with the argument `edit`
this succeeds, but when I try to build the project, I get an error:
```
Showing Recent Messages
~/Desktop/Scratch/Tuist/ProjectDescriptionHelpers/Project+Templates.swift:1:8: No such module 'ProjectDescription'

```
Likewise, if I run `tuist generate`, I get
```
~/Desktop/Scratch/Tuist/ProjectDescriptionHelpers/Project+Templates.swift:1:8: error: no such module 'ProjectDescription'
import ProjectDescription
       ^
[31mThe 'swiftc' command exited with error code 1[0m
Program ended with exit code: 1
```

> Some final bits _(thanks for your patience!)_:

I appreciate your help putting this together. Thank _you_ for your patience.

> * Looks like the latest merge from master may have gone a bit funny, the PR is showing unrelated changes included :(

Sigh. I'll take a peek.

> * You can run `bin/swiftformat Sources Tests` to ensure the same version that is used on CI formats the code

Will do.

> Minor suggestions:
> 
> * The `TargetActionEnvironmentMapper`/Tests could be in its own file in `TuistGenerator/Mappers` rather that part of the `TargetMapper` file in the core library
> * `ConfigTests` can be extended to also include `disableShowEnvironmentVarsInScriptPhases`

I'll do this as well.

kalkwarf commented:
> > Were you able to get it up and running locally with the latest updates?
> No, I'm really fighting Xcode & SPM.

It appears that the problem is related to search paths. In the project generated by `tuist edit`, the framework search path points to `~/Library/Developer/Xcode/DerivedData/tuist-bggkhtdqqctkgpgksefefjesrvfj/Build/Products/Debug/PackageFrameworks`, and in that directory the `ProjectDescription.framework` does exist. However, `ProjectDescription.swiftmodule` is in the parent directory (`Debug`). 

Changing the three search paths to point at `Debug` allows the Manifests project to build.

kwridan commented:
Ouch looks like some more changes are needed for to support `tuist edit` via Xcode - we could look into that separately (I've added #1814 for this). 

For the moment, you can skip the `tuist edit` step via Xcode, the `tuist generate` command should still work in Xcode.

An alternative is running commands from the command line via `swift run tuist <subcommand>` 

e.g.

```sh
cd path/to/tuist
cd fixtures/ios_app_with_tests
swift build
swift run tuist edit
swift run tuist generate
# ...
```

kwridan commented on fixtures/ios_app_with_env_suppressed/App With Space/Project.swift:
```diff
@@ -0,0 +1,14 @@
+import ProjectDescription
```
👍 thanks for adding the fixture! This will help us easily test this feature

kwridan reviewed (commented):
A few final tweaks @kalkwarf - thanks for all the updates

When its ready, please update the `CHANGELOG`

kalkwarf commented on Tests/TuistCoreTests/Graph/Mappers/ConfigShowEnvironmentTests.swift:
```diff
@@ -0,0 +1,47 @@
+import TuistCore
+import TuistCoreTesting
+@testable import TuistGenerator
```
Ironically, this test isn't valid any more. It was left over from when I was setting the flag at the project level, where now it's handled at Config.

kwridan replied:
The test is/was still valid :) 

It's testing the `BuildPhaseGenerator` component respects the `TargetAction.showEnvVarsInLog` flag

The current flow is

- `Config`: Enables `TargetActionEnvironmentMapper`
- `TargetActionEnvironmentMapper`: sets the `TargetAction.showEnvVarsInLog` to `false`
- `BuildPhaseGenerator`: Consumes `Target` and `TargetAction`s to generate the Xcode build phases

That test covered that last bit.

kalkwarf commented on Tests/TuistCoreTests/Graph/Mappers/ProjectMapperTests.swift:
```diff
@@ -54,6 +54,31 @@ final class TargetProjectMapperTests: XCTestCase {
         ])
     }
 
+    func test_map_environmentLoggingToggles() throws {
```
That makes sense. Sorry about that; I'm still learning my way around.

Again, thanks for the handholding.

kwridan commented on Sources/TuistCore/Graph/Mappers/TargetMapper.swift:
```diff
@@ -3,3 +3,22 @@ import Foundation
     func map(target: Target) throws -> (Target, [SideEffectDescriptor])
 }
+
+public class TargetActionEnvironmentMapper: TargetMapping {
```
minor: The implementation can move too https://github.com/tuist/tuist/pull/1782#discussion_r493557151

 but happy to do this afterwards too - you've suffered my feedback long enough :p

kalkwarf replied:
I have nothing but appreciation for your help. I'll move this over, too.

I have a personal aversion to "Can I fix this in a follow-up PR?" because in my experience, the follow-up rarely materializes.

kwridan reviewed (approved):
Awesome - thanks for all the updates and patience on this one, once it's CI is ✅ - its good to 🚀

fortmarek commented on Sources/TuistGenerator/Mappers/TargetActionEnvironmentMapper.swift:
```diff
@@ -0,0 +1,20 @@
+import TuistCore
+
+public class TargetActionEnvironmentMapper: TargetMapping {
```
Nitpick: I'd mark this `final`
I'd also add just a quick documentation that explains what this mapper is for

kalkwarf replied:
Done

fortmarek commented on Sources/TuistKit/GraphMappers/ProjectMapperProvider.swift:
```diff
@@ -29,6 +29,13 @@ final class ProjectMapperProvider: ProjectMapperProviding {
         }
 
+        // Logfile noise suppression
+        if config.generationOptions.contains(.disableShowEnvironmentVarsInScriptPhases) {
```
I'd add a simple test case similar to `test_mappers_does_not_returns_resources_namespace_project_mapper_when_disabled_autogenerated_namespace` in `ProjectMapperProviderTests`

kalkwarf replied:
Done

fortmarek commented on Tests/TuistGeneratorTests/Mappers/ConfigShowEnvironmentMapperTests.swift:
```diff
@@ -0,0 +1,56 @@
+@testable import TuistGenerator
+@testable import TuistSupportTesting
+
+public final class ConfigShowEnvironmentMapperTests: TuistUnitTestCase {
```
hmm, does this need to be `public`?

kalkwarf replied:
Nope. Changed.

fortmarek reviewed (approved):
Just small notes, I see that you have had great guidance from @kwridan, so he's covered most of the ground already 🙂

fortmarek commented:
LGTM; I'll just let @kwridan take a quick last look and let him merge this 🙂

kwridan commented:
Awesome work @kalkwarf 👏

kalkwarf commented:
> Awesome work @kalkwarf 👏

Thanks for all the guidance. Much appreciated.
